### DISCUSSION TITLE
Interactive CLI

### DISCUSSION THREAD

**Discussion by morpheuslord** [community/community#62550] [Programming Help] [answered]
### Body

So In my latest projects, I wanted to implement a CLI interface similar to htop, btop and a few others. but in the process, I considered using the curses module which seemed promising but super hard to work with and understand are there any other useful modules for an interactive and responsive CLI experience while having good looks and feel

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by ViridianTelamon** [comment]
Honestly, for applications in the terminal that module that you mentioned is the best one, that module has many features and is extremely useful, but while I will suggest to try to use it one more time, if you really want you could use Kivy or PyQt to make a program that is like that.

**Reply by morpheuslord** [reply]
Kivy and PyQt both are for GUI applications and not the CLI interactive one I was searching for. But I will try curses and maybe make it work.

**Reply by ViridianTelamon** [reply]
It does usually work well for CLI applications and it is usually also the best for CLI applications, maybe check out the Tech With Tim YouTube channel as he has videos on it, and maybe you should also check the documentation for it as well.

**Reply by morpheuslord** [reply]
Thanks for the resource I will look into it

**Reply by ViridianTelamon** [reply]
np

**Comment by Jaber-Saed** [comment, marked as answer]
Certainly, the `curses` module in Python is powerful but can indeed be complex to work with. However, there are other libraries that can help you create interactive and responsive CLI applications with a good user experience. Here are a few alternatives you might consider:

1. **Click**: Click is a popular library for creating command-line interfaces with a clean and user-friendly syntax. While it's not meant for creating terminal UIs like `htop`, it can help you create clear and structured command-line tools that are easy to use.

2. **Prompt Toolkit**: This library provides a high-level interface for building interactive command-line applications. It includes widgets like text input, menus, checkboxes, and more. It's less complex than `curses` but still offers a rich set of features for building interactive CLI apps.

3. **Rich**: Rich is a library for adding color and style to terminal output. It doesn't focus on interactive widgets like `curses`, but it can enhance the look and feel of your CLI applications by providing text formatting, tables, progress bars, and more.

4. **Inquirer**: Inquirer is a Python implementation of the Inquirer.js library, which is used to create interactive command-line prompts. It's suitable for creating interactive questionnaires and prompts for user input.

5. **asciimatics**: This is another library that provides an easy way to create interactive terminal applications using a simple API. It's designed to make it easier to create TUI (Text User Interface) applications with fewer complexities than `curses`.

6. **textual**: Textual is a library built on top of `curses` that aims to simplify the process of creating terminal applications with a TUI. It provides a more modern and intuitive API compared to raw `curses` programming.

Each of these libraries serves a slightly different purpose, so the choice depends on the complexity of your CLI application and the level of interaction you need to provide to your users. Remember that, while these libraries can simplify the development process, creating a polished and well-designed CLI application still requires careful consideration of the user experience and design principles.

**Reply by morpheuslord** [reply]
Thanks a lot for the info I will certainly look into it.
